census/vanilla-tf/trainer/task.py

```diff
@@ -0,0 +1,349 @@
+  dense_tensor = tf.cast(dense_tensor, tf.int32)
+  return dense_tensor
+
+def read_input_tensor(input_file, skiprows=None):
+  """Concatenate the wide columns to produce a single tensor."""
+  inp, label = read_input_data(input_file, skiprows)
+  in_tensor, label_tensor = generate_input(inp, label)
+  return concat_wide_columns(generate_wide_columns(in_tensor)), label_tensor
+
+
+#
+# Function to perform the actual training loop.
+# This function is same for single and distributed.
+#
+def training_steps(session, train_step, eval_step, model, labels, global_step, max_steps,
+                   inp, label,
+                   eval_inp, eval_label):
+  """Run the training steps and calculate accuracy every 10 steps."""
+  step = tf.train.global_step(session, global_step)
+
+  while step < max_steps:
+    session.run(
+        train_step,
+        feed_dict={
+            inputs: session.run(inp),
+            labels: session.run(label)
+            }
+    )
+
+    step = tf.train.global_step(session, global_step)
+
+    if step % 10 == 0:
+      accuracy = evaluate_accuracy(session, eval_step, eval_inp, eval_label)
+      print('Step number {} of {} done, Accuracy {:.2f}%'.format(
+          step, max_steps, accuracy))
+
+  return train_step
+
+
+#
+# Single and Distributed training functions.
+#
+def training_single(model, labels, max_steps,
+                    inp_tensor, label_tensor,
+                    eval_inp_tensor, eval_label_tensor):
+  """Perform single node training."""
+
+  global_step = tf.contrib.framework.get_or_create_global_step()
+  init = tf.global_variables_initializer()
+
+  session = tf.Session()
+  session.run(init)
+
+  cross_entropy = loss(model, labels)
+  train_step = optimizer(cross_entropy, global_step)
+  accuracy = evaluation_step(model, labels)
+
+  training_steps(session, train_step, accuracy, model, labels, global_step, max_steps,
+                 inp_tensor, label_tensor,
+                 eval_inp_tensor, eval_label_tensor)
+
+  return train_step
+
+def training_distributed(model, labels, max_steps,
+                         inp_tensor, label_tensor,
+                         eval_inp_tensor, eval_label_tensor):
+  """Perform distributed training."""
+
+  cluster_spec, job_name, task_index = parse_tf_config()
+
+  # Create and start a server
+  server = tf.train.Server(cluster_spec,
+                           job_name=job_name,
+                           task_index=task_index)
+
+  is_chief = (job_name == 'master')
+
+  if job_name == 'ps':
+    server.join()
+  elif job_name in ['master', 'worker']:
+    with tf.device(tf.train.replica_device_setter()):
+      global_step = tf.contrib.framework.get_or_create_global_step()
+
+      cross_entropy = loss(model, labels)
+      train_step = optimizer(cross_entropy, global_step)
+      accuracy = evaluation_step(model, labels)
+
+    init = tf.global_variables_initializer()
+
+    with tf.train.MonitoredTrainingSession(master='', is_chief=is_chief) as session:
```

Comment:
You need  `master=server.target` to do distributed properly. 

FYI you can also use a MonitoredTrainingSession for local if you want to, and just use `master=''` that might help with code reuse.